chain/network/src/concurrency/scope/mod.rs

```diff
@@ -0,0 +1,444 @@
+//! Asynchronous scope.
+//! WARNING: this module assumes that [panic=abort] is enabled.
+//!
+//! You can think of the scope as a lifetime 'env within a rust future, such that:
+//! * within 'env you can spawn subtasks, which may return an error E.
+//! * subtasks can spawn more subtasks.
+//! * 'env has special semantics: at the end of 'env all spawned substasks are awaited for
+//!    completion.
+//! * if ANY of the subtasks returns an error, all the other subtasks are GRACEFULLY cancelled.
+//!   It means that they are not just dropped, but rather they have a handle (aka Ctx) to be able
+//!   to check at any time whether they were cancelled.
+//!
+//!     let (send,recv) = channel();
+//!     ...
+//!     'env: {
+//!         spawn<'env>(async {
+//!             recv.await
+//!             Err(e)
+//!         });
+//!
+//!         for !is_cancelled() {
+//!             // do some useful async unit of work
+//!         }
+//!         // do some graceful cleanup.
+//!         Ok(())
+//!     }
+//!
+//! Since we cannot directly address lifetimes like that we simulate it via Scope and Ctx structs.
+//! Ctx is hidden in the thread_local storage and is configured by Scope.
+//! We cannot directly implement a function
+//!   run : (Scope<'env> -> (impl 'env + Future)) -> (impl 'env + Future)
+//! Because the compiler is not smart enough to deduce 'env for us.
+//! Instead we first construct Scope<'env> explicitly, therefore fixing its lifetime,
+//! and only then we pass a reference to it to another function.
+//!
+//!     let (send,recv) = channel();
+//!     ...
+//!     {
+//!         let s = Scope<'env>::new();
+//!         s.run(ctx,|s| async {
+//!             s.spawn(async {
+//!                 recv.await
+//!                 Err(e)
+//!             })
+//!
+//!             for !ctx::is_cancelled() {
+//!                 // do some useful async unit of work
+//!             }
+//!             // do some graceful cleanup.
+//!             Ok(())
+//!         }).await
+//!     }
+//!
+//! We wrap these 2 steps into a macro "run!" to hide this hack and avoid incorrect use.
```

Comment:
I’m curious why a function wouldn’t work when it does for std::thread::scope; but I guess that’ll become clear when I’ll actually read the code.

Reply:
`std::thread::scope` is taking a function of type `Scope -> T`, where T has static lifetime. Here we take a function which returns a future of lifetime depending on the lifetime of the argument. That's where the compiler gives up when determining the type of the closure.